course2/course_en.md

```diff
@@ -0,0 +1,555 @@
+  - Conversion from `Double` to `Int`: `(-1.2).to_int() == -1`
+- Quiz: How to compare `0.1 + 0.2` with `0.3` using `Int` and `Double` conversions?
+
+---
+
+# Characters and Strings
+
+- In MoonBit, the character type is `Char`, and the string type is `String`.
+    - Characters are represented with single quotes: `'a'`
+    - Strings are represented with double quotes: `"Hello!"`
+- Characters have different encoding schemes.
+    - ASCII (American Standard Code for Information Interchange): Ranges from 0 to 127, supporting Latin alphabet, Arabic numerals, and common symbols.
+        - For example, 'A' to 'Z' corresponds to 65 to 90.
+    - Unicode: Supports characters from various languages, including Chinese characters and emoji. It is compatible with ASCII and has different variants like UTF-8 and UTF-16.
+        - For example, "月" and "兔" correspond to 26376 and 20820, respectively.
+    - Converting an `Int` to a `Char`: `Char.fromInt(65) == 'A'`.
+
+---
+
+# Tuples
+
+- Tuples allow us to combine data of different types with a fixed length.
+  - `("Bob", 3): (String, Int)`
+  - `(2023, 10, 24): (Int, Int, Int)`
+- We can access the data by using zero-based indexing.
+  - `(2023, 10, 24).0 == 2023`
+  - `(2023, 10, 24).1 == 10`
+
+---
+
+# Other Data Types
+
+- MoonBit has a rich type system.
+  - Functions
+    - `op_add: (Int, Int) -> Int`
+  - Unit type (representing no value)
+    - `(): Unit`
+  - Lists
+    - `List.Cons(1, Nil): List<Int>`
+  - ...
+- We will see more about these types and custom data structures in future lessons.
+
+---
+
+# Expression Evaluation
+
+---
+
+# Reduction vs Execution
+
+- MoonBit expressions can be seen as a way of representing **values**.
+- Its evaluation can be seen as a series of **computations** or **reductions**.
+- In contrast, imperative programming can be seen as executing a series of **actions** or **commands**, where each command modifies the state of the machine.
+    - Create pointers `x` and `y` and allocate memory, set `x` to 3, set `y` to 4.
+    - Set `y` to point to `x`.
+    - Increment `x`.
+    - ...
+    
+![](../pics/local-store.drawio.png)
+
+---
+
+# Expression Reduction
+
+- $3 \Rightarrow 3$ (the reduction result of a value is itself)
+- $3 + 4 \Rightarrow 7$
+- $2 * (4 + 5) \Rightarrow 18$
+- $\mathtt{num\_water\_bottles}(9, 3) \Rightarrow 13$
+
+We denote the reduction of an $\texttt{<expression>}$ to a $\texttt{<value>}$ as:
+- $\texttt{<expression>} \Rightarrow \texttt{<value>}$
+
+---
+
+# Step-by-Step Computation
+
+We break down the process of decomposition $\Rightarrow$ reduction into **step-by-step** computation, denoted as $\mapsto$.
+
+For example:
+$$\begin{align}
+    & (2 + 3) * (5 - 2) \\
+    \mapsto & 5 * (5 - 2) && \text{because $2 + 3 \mapsto 5$}\\
+    \mapsto & 5 * 3 && \text{because $5 - 2 \mapsto 3$} \\
+    \mapsto & 15 && \text{because $5 * 3 \mapsto 15$}
+\end{align}$$
+
+Therefore, $(2 + 3) * (5 - 2) \Rightarrow 15$.
+
+---
+
+# Variable Binding
```

Comment:
I use "Variable Binding" instead of "Value Binding" to conform to the terminology in the MoonBit Docs, the same below.